src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    internal static partial class JsonReaderHelper
+    {
+        // Reject any invalid UTF-8 data rather than silently replacing.
+        public static readonly UTF8Encoding s_utf8Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
+
+        private const int StackallocThreshold = 256;
+
+        // TODO: Similar to escaping, replace the unescaping logic with publicly shipping APIs from https://github.com/dotnet/corefx/issues/33509
+        public static string GetUnescapedString(ReadOnlySpan<byte> utf8Source, int idx)
+        {
+            byte[] unescapedArray = null;
+            string utf8String;
+
+            Span<byte> unescaped = utf8Source.Length <= StackallocThreshold ?
+                stackalloc byte[utf8Source.Length] :
+                (unescapedArray = ArrayPool<byte>.Shared.Rent(utf8Source.Length));
+
+            Unescape(utf8Source, unescaped, idx, out int written);
+            utf8String = s_utf8Encoding.GetString(unescaped.Slice(0, written));
+
+            if (unescapedArray != null)
+            {
+                ArrayPool<byte>.Shared.Return(unescapedArray);
```

Comment:
We've been inconsistent across our libraries in general, but even just our new JSON support, about whether arrays rented from the pool are cleared before being returned.  Can you sync with @bartonjs on what we want to do here?  When the arrays contain user data, IIRC Jeremy's been clearing them.  I don't have a strong opinion about which way we should go, other than we should be consistent, and either always clear in such situations or never clear in such situations.

Reply:
The JSON stuff needs to always clear user data, because it's expected to be used in multi-threaded server scenarios, and it needs to not be a bad steward of that data.  (This is but one piece in a set of dominos for a user data leak, but historically (due to a historical inability to get "used" memory) one that was not possible in .NET... we need to not regress the security benefits)